### PR TITLE
feat(eks): IAM roles for service accounts

### LINKED ISSUE #3949: aws-eks with IRSA(IAM Roles for Service Accounts) support
### General Information

* [ ] :wave: I may be able to implement this feature request

* [ ] :warning: This feature might incur a breaking change

### Description

AWS just announced the **IRSA(IAM Roles for Service Accounts)** support for Amazon EKS as well as DIY Kubernetes on AWS.
https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/
https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html

It would be great if we can optionally specify the pod role as a property for the [aws-eks/lib/k8s-resource](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-eks/lib/k8s-resource.ts) to simplify the heavy-lifting.

### Environment

  - **CDK CLI Version:** 1.6.1
  - **Module Version:** 1.6.1
  - **OS:** all
  - **Language:** all

### LINKED ISSUE #5388: EKS OpenID Connect provider association with AWS account
Freature refers to fine-grained IAM roles for service accounts. In this approach, pods are pods first class citizens in IAM. To make the feature work, OpenID provider which comes with EKS needs to be associated with AWS account. More one here: https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html

eksctl does that in one command. Unfortunatelly there is no native cloudformation support so it cannot be set up automatically.

The goal would be to add a parameter to EKS cluster constructor, which would let programmer decide whether to associate the EKS OpenID with AWS IAM. 

### Use Case

I'd like to use AWS native support for IAM roles for pods rather than 3rd party solutions like role-credential interceptors (kube2iam for instance).

### Proposed Solution

kubectl_enabled flag (introduced in CDK 1.4, if I recall correctly), changes the approach to the way how EKS is provisioned. You could follow the same route and perform custom logic after cluster is deployed. The logic would need to perform following steps (more about this here: https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html):

- retrieve OID provider URL from the cluster
- create open id identity provider (https://docs.aws.amazon.com/cli/latest/reference/iam/create-open-id-connect-provider.html). The only difficulty is to get the thumbprint (more about this here: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc_verify-thumbprint.html)

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by vlesierse**
## Commit Message
feat(eks): IAM roles for service accounts (#6062)

Adds support for IAM roles for service account which allows pods the assume IAM roles.

NOTE: currently there are no condition set on the IAM Role which results that there are no restrictions on other pods to assume the role. This will be fixed in a subsequent PR.

See README for details.

Fixes #5388 
Fixes #3949 

## End of Commit Message

- [x] Enable OpenID Connect Provider
- [x] Service Account construct
- [ ] Role constraints
- [x] Add `cluster.addServiceAccount` convenience method
- [x] Integration Tests
- [x] Unit Tests
- [x] Update README.md

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by vlesierse** [comment]
@eladb question regarding CDK as related to this PR.

Following to the [EKS Documentation](https://docs.aws.amazon.com/eks/latest/userguide/create-service-account-iam-policy-and-role.html) I'm trying create an IAM role with an assume policy which contains a condition. This is something that can easily be done with CDK. However the condition's key needs to be a calculated value of the custom resource, which is a token in CDK. CloudFormation doesn't support this. Is there are way to resolve this before passing it through?

```typescript
const condition: { [id: string]: any; } = {};
condition[this.openIDConnectSubject!] = `system:serviceaccount:${namespace}:${name}`;
```

**Comment by eladb** [comment]
> CloudFormation doesn't support this. Is there are way to resolve this before passing it through?

Yes, indeed, CloudFormation doesn't support using intrinsic functions as map keys... Can you provide a bit more context into your code? I am not sure I fully see where you are using this.

**Comment by vlesierse** [comment]
I've pushed the code how it should work as you can see in the `service-account.ts`. My alternative is to create Customer Resource to swap the condition, but it feels a bit heavy for something this small..

**Comment by vlesierse** [comment]
@eladb the initial implementation is working. Eventually I've solved the problem if IAM condition keys with a CustomerResource. It's not the most elegant solution, but I think the only one available at the moment.

I'm currently focusing on the tests.

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcprovider.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        openIDConnectProviderArn: createOpenIDConnectProviderResponse.OpenIDConnectProviderArn
+      }
+    };
+  }
+
+  protected async onDelete() {
+    if (!this.physicalResourceId) {
+      throw new Error(
+        `Cannot delete a profile without a physical id`
+      );
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcprovider.ts]:
```suggestion
      throw new Error(`Cannot delete an OIDC provider without a physical id`);
```

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcprovider.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    const deleteOpenIDConnectProvider: aws.IAM.DeleteOpenIDConnectProviderRequest = {
+      OpenIDConnectProviderArn: this.physicalResourceId
+    };
+
+    this.log({ deleteOpenIDConnectProvider });
+    const deleteOpenIDConnectProviderResponse = await this.eks.deleteOpenIDConnectProvider(
+      deleteOpenIDConnectProvider
+    );
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcprovider.ts]: style nit:

```suggestion
    const deleteOpenIDConnectProviderResponse = await this.eks.deleteOpenIDConnectProvider(deleteOpenIDConnectProvider);
```

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcprovider.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+      this.log({ getOpenIDConnectProvider });
+      const getOpenIDConnectProviderResponse = await this.eks.getOpenIDConnectProvider(
+        getOpenIDConnectProvider
+      );
+      this.log({ getOpenIDConnectProviderResponse });
+      return true;
+    } catch (getOpenIDConnectProviderError) {
+      if (getOpenIDConnectProviderError.code === "NoSuchEntity") {
+        this.log(
+          "received NoSuchEntityFoundException, this means the profile has been deleted (or never existed)"
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcprovider.ts]:
```suggestion
          "received NoSuchEntityFoundException, this means the provider has been deleted (or never existed)"
```

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcrole.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    const document = JSON.parse(decodeURIComponent(role.Role.AssumeRolePolicyDocument!));
+    const statement = (document.Statement as any[]).find(s => s.Effect === "Allow" && s.Action === "sts:AssumeRoleWithWebIdentity");
+
+    const condition: { [id: string]: any} = {};
+    condition[`${issuerUrl}:aud`] = "sts.amazonaws.com";
+    condition[`${issuerUrl}:sub`] = `system:serviceaccount:${serviceAccountNamespace}:${serviceAccountName}`;
+    if (!statement.Condition) {
+      statement.Condition = {};
+    }
+    statement.Condition.StringEquals = condition;
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcrole.ts]: what happens if the policy already has a `statement.Condition.StringEquals`?

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcrole.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  protected async onDelete() {
+    if (!this.physicalResourceId) {
+      throw new Error(
+        `Cannot get a role without a physical id`
+      );
+    }
+
+    const role = await this.eks.getRole({ RoleName: this.physicalResourceId});
+    if (!role) {
+      throw new Error(`no role found with name ${this.physicalResourceId}`);
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcrole.ts]: Should we just consider this a success? (idempotency)

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcprovider.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    if (!this.physicalResourceId) {
+      throw new Error(
+        `Cannot delete a profile without a physical id`
+      );
+    }
+
+    const deleteOpenIDConnectProvider: aws.IAM.DeleteOpenIDConnectProviderRequest = {
+      OpenIDConnectProviderArn: this.physicalResourceId
+    };
+
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/oidcprovider.ts]: Maybe call "checkResource" first and return success if it's not found (idempotency)

**Comment by RafalWilinski** [comment]
Do you need any help with this PR? Happy to contribute some code in order to push this forward.

**Comment by vlesierse** [comment]
Thanks @RafalWilinski your help would be appreciated. The coming week I might have some more time to finish this. The code is working but needs a bit of refactoring and the addition of tests.

@eladb I'm wondering if the code needs to move the to `aws-iam` as the logic is pretty EKS specific. Updating the assume role policy might be.

**Comment by eladb** [comment]
> @eladb I'm wondering if the code needs to move the to aws-iam as the logic is pretty EKS specific. Updating the assume role policy might be.

Yes, I think it needs to move to IAM, which makes this dramatically more complex. Let me pick this task up and then you'll be able to leverage this work here. OK?

**Comment by vlesierse** [comment]
@eladb sure, that will be great. Last night I was struggling a bit how to put this in `aws-iam`. I will focus today on some of the testing and documentation and adjust the code once you are done.

**Comment by vlesierse** [comment]
> Would be nice to validate this logic against https://github.com/smaato/cfn-resource-provider.

I have used the same logic as what they regarding creating the provider. However their update logic is more sophisticated which I can change in my implementation to be more resilient to to changes from EKS. (Client ID or Certificate changes)

**Comment by eladb** [comment]
OpenID connect provider resource is coming up... Stay tuned.

**Comment by vlesierse** [comment]
@eladb great news! Is there any branch available where I can take a look?

**Comment by eladb** [comment]
> @eladb great news! Is there any branch available where I can take a look?

https://github.com/aws/aws-cdk/pull/7803

File: packages/@aws-cdk/aws-eks/test/cluster-resource-handler-mocks.ts [hunk 7]
```diff
@@ -136,6 +136,7 @@ export const client: EksClient = {
     actualRequest.deleteFargateProfile = req;
     return { };
   },
+
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/test/cluster-resource-handler-mocks.ts:139]:
```suggestion
```

File: packages/@aws-cdk/aws-eks/lib/cluster-resource.ts [hunk 8]
```diff
@@ -99,16 +100,6 @@ export class ClusterResource extends Construct {
       resources: [ fargateProfileResourceArn ],
     }));
 
-    this.creationRole.addToPolicy(new iam.PolicyStatement({
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource.ts:102]: why were these deleted (probably a bad merge)?

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+   * The role the service account is linked to.
+   */
+  public readonly role: IRole;
+
+  constructor(scope: Construct, id: string, props: ServiceAccountProps) {
+    super(scope, id);
+
+    const { cluster, name, namespace } = props;
+
+    let provider = cluster.node.tryFindChild('OpenIdConnectProvider') as OpenIdConnectProvider;
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:45]: I rather move this to the Cluster class as `cluster.openIdConnectProvider` as a lazy-initialized property:

```ts
class Cluster {
  public get openIdConnectProvider() {
    if (!this._openIdConnectProvider) {
      if (!this.kubectlEnabled) { throw new Error(`...`); }
      this._openIdConnectProvider = new OpenIdConnectProvider(this, 'OpenIdConnectProvider', {
        url: this.clusterOpenIdConnectIssuerUrl
      }
    }
    return this._openIdConnectProvider;
  }
}
```

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    if (!provider) {
+      if (!cluster.kubectlEnabled) {
+        throw new Error('Cannot specify a OpenID Connect Provider if kubectl is disabled');
+      }
+      provider = new OpenIdConnectProvider(cluster, 'OpenIdConnectProvider', {
+        url: props.cluster.clusterOpenIdConnectIssuerUrl!,
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:51]: Why is the `!` needed? This will always have a value

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 11]
```diff
@@ -298,6 +298,12 @@ export class Cluster extends Resource implements ICluster {
    */
   public readonly clusterCertificateAuthorityData: string;
 
+  /**
+   * The OpenID Connect Issuer Url for your cluster.
+   * @attribute
+   */
+  public readonly clusterOpenIdConnectIssuerUrl?: string;
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:305]: Why is this optional (attributes are always defined)?

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      });
+    }
+
+    this.role = new Role(this, 'Role', {
+      assumedBy: new FederatedPrincipal(provider.openIdConnectProviderArn, {}, 'sts:AssumeRoleWithWebIdentity'),
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:56]: Should we add an `iam.WebIdentityPrincipal` that will wrap this? Alternatively, we can me `OpenIdConnectProvider` an `iam.IPrincipal` or add an `openIdConnectProvider.principal`

File: packages/@aws-cdk/aws-eks/package.json [hunk 13]
```diff
@@ -63,6 +63,7 @@
   "license": "Apache-2.0",
   "devDependencies": {
     "@aws-cdk/assert": "0.0.0",
+    "@types/node": "^10.17.17",
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/package.json:66]: I think this is installed globally and not needed in each module

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 11 above)
Comment (vlesierse) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:305]: The CfnCluster resource doesn't expose this attribute. I can only resolved using the API. This is why kubectlEnabled must be true. Only then it can get the OpenIdConnectProvider.
An alternative would be to use a custom resource for just retrieving the Issuer URL.

File: packages/@aws-cdk/aws-eks/lib/service-account.ts (hunk 12 above)
Comment (vlesierse) [reply packages/@aws-cdk/aws-eks/lib/service-account.ts:56]: It's actually a common pattern and usage with IAM. Only the condition would vary but it's always a FederatedPrincipal with action sts:AssumeRoleWithWebIdentity.

Might it be a solutions as well to add the conditions using a custom resource to overcome the lack of support for intrinsic functions as map keys? The conditions are important as security feature. Without it allows every pod to assume the role. The condition pervents this.

File: packages/@aws-cdk/aws-eks/lib/service-account.ts (hunk 10 above)
Comment (vlesierse) [reply packages/@aws-cdk/aws-eks/lib/service-account.ts:51]: If `kubectlEnabled` is true you can assume that the property has a value.

Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/service-account.ts:51]: It should always have a value

File: packages/@aws-cdk/aws-eks/lib/service-account.ts (hunk 12 above)
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/service-account.ts]: Yes, definitely sounds like heavy lifting we can offer. Do you want to take a whack at it?

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/consts.ts [hunk 14]
```diff
@@ -1,2 +1,4 @@
 export const CLUSTER_RESOURCE_TYPE = 'Custom::AWSCDK-EKS-Cluster';
-export const FARGATE_PROFILE_RESOURCE_TYPE = 'Custom::AWSCDK-EKS-FargateProfile';
\ No newline at end of file
+export const FARGATE_PROFILE_RESOURCE_TYPE = 'Custom::AWSCDK-EKS-FargateProfile';
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/consts.ts:2]: revert

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 15]
```diff
@@ -617,6 +628,23 @@ export class Cluster extends Resource implements ICluster {
...
+   * Lazily creates the OpenIdConnectProvider resource, which links the cluster to AWS IAM.
+   */
+  public get openIdConnectProvider() {
+    if (!this.kubectlEnabled) {
+      throw new Error('Cannot specify a OpenID Connect Provider if kubectl is disabled');
+    }
+
+    if (!this._openIdConnectProvider) {
+      this._openIdConnectProvider = new iam.OpenIdConnectProvider(this, 'OpenIdConnectProvider', {
+        url: this._clusterOpenIdConnectIssuerUrl!,
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:641]: Now you can safely do this:

```suggestion
        url: this.clusterOpenIdConnectIssuerUrl,
```

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 16]
```diff
@@ -337,11 +337,21 @@ export class Cluster extends Resource implements ICluster {
    */
   private readonly _clusterResource?: ClusterResource;
 
+  /**
+   * If this cluster is kubectl-enabled, returns the OpenID Connect issuer url.
+   * This is because the values is only be retrieved by the API and not exposed
+   * by CloudFormation. If this cluster is not kubectl-enabled (i.e. uses the
+   * stock `CfnCluster`), this is `undefined`.
+   */
+  private readonly _clusterOpenIdConnectIssuerUrl?: string;
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:346]: Replace this with:

```ts
public get clusterOpenIdConnectIssuerUrl(): string {
  if (!this._clusterResource) {
    throw new Error(`unable to obtain OpenID Connect issuer URL. Cluster must be kubectl-enabled`);
  }

  return this._clusterResource.attrOpenIdConnectIssuerUrl;
}
```

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 17]
```diff
@@ -402,6 +412,7 @@ export class Cluster extends Resource implements ICluster {
     if (this.kubectlEnabled) {
       resource = new ClusterResource(this, 'Resource', clusterProps);
       this._clusterResource = resource;
+      this._clusterOpenIdConnectIssuerUrl = resource.attrOpenIdConnectIssuerUrl;
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:415]: This is not needed

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+ */
+export interface ServiceAccountOptions {
+  /**
+   * The cluster to apply the patch to.
+   * @default If no name is given, it will use the id of the resource.
+   */
+  readonly name?: string;
+
+  /**
+   * The cluster to apply the patch to.
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:16]:
```suggestion
   * The namespace to apply the patch to.
```

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+   * The namespace where the service account is located in.
+   */
+  public readonly serviceAccountNamespace: string;
+
+  constructor(scope: Construct, id: string, props: ServiceAccountProps) {
+    super(scope, id);
+
+    const { cluster } = props;
+    this.serviceAccountName = props.name || id;
+    this.serviceAccountNamespace = props.namespace || 'default';
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:62]:
```suggestion
    this.serviceAccountNamespace = props.namespace ?? 'default';
```

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+  /**
+   * The namespace where the service account is located in.
+   */
+  public readonly serviceAccountNamespace: string;
+
+  constructor(scope: Construct, id: string, props: ServiceAccountProps) {
+    super(scope, id);
+
+    const { cluster } = props;
+    this.serviceAccountName = props.name || id;
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:61]:
```suggestion
    this.serviceAccountName = props.name ?? id;
```

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+    this.serviceAccountNamespace = props.namespace || 'default';
+
+    this.role = new Role(this, 'Role', {
+      assumedBy: new FederatedPrincipal(props.cluster.openIdConnectProvider.openIdConnectProviderArn, {}, 'sts:AssumeRoleWithWebIdentity'),
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:65]:
```suggestion
      assumedBy: new FederatedPrincipal(cluster.openIdConnectProvider.openIdConnectProviderArn, {}, 'sts:AssumeRoleWithWebIdentity'),
```

File: packages/@aws-cdk/aws-eks/lib/service-account.ts (hunk 12 above)
Comment (vlesierse) [reply packages/@aws-cdk/aws-eks/lib/service-account.ts]: @eladb I'm busy with adding the WebIdentityPrincial which make is a little bit easier to work with. However I'm still struggling a bit the dynamic map key of the EKS version of the WebIdentityPrincipal. I can't pass the value directly from the cluster as it is an attribute reference instead of the real value. This means I have two options to pursuit:

1. Create a dedicated custom resource to change the condition based on what is given by the ServiceAccount construct in either the iam module or eks.
2. Create a policy document template custom resource in iam which allows you to replace `${Foo}` with `Bar` in a policy document. 

What I noticed in the documentation that there are 4 common ones (Cognito, Amazon, Facebook & Google) which we can create a class off and add the conditions using properties. But this is something for a separate PR.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 22]
```diff
@@ -617,6 +620,45 @@ export class Cluster extends Resource implements ICluster {
...
+    if (!this._openIdConnectProvider) {
+      this._openIdConnectProvider = new iam.OpenIdConnectProvider(this, 'OpenIdConnectProvider', {
+        url: this.clusterOpenIdConnectIssuerUrl,
+        clientIds: [ 'sts.amazonaws.com' ],
+        /**
+         * For some reason EKS isn't validating the root certificate but a intermediat certificate
+         * which is one level up in the tree. Because of the a constant thumbprint value has to be
+         * stated with this OpenID Connect provider. The certificate thumbprint is the same for all the regions.
+         */
+        thumbprints: [ '9e99a48a9960b14926bb7f3b02e22da2b0ab7280' ],
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:655]: ok....

File: packages/@aws-cdk/aws-iam/lib/principals.ts [hunk 23]
```diff
@@ -382,6 +382,31 @@ export class FederatedPrincipal extends PrincipalBase {
...
+ */
+export class WebIdentityPrincipal extends FederatedPrincipal {
+
+  /**
+   *
+   * @param identityProvider identity provider (i.e. 'cognito-identity.amazonaws.com' for users authenticated through Cognito)
+   * @param conditions The conditions under which the policy is in effect.
+   *   See [the IAM documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition.html).
+   */
+  constructor(identityProvider: string, conditions: Conditions = {}) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/lib/principals.ts:397]: Would be nice to make this a not a `string` or maybe just add subclasses such as `OpenIdPrincipal` that extends this an accepts an `IOpenIdProvider` object.

File: packages/@aws-cdk/aws-eks/README.md [hunk 24]
```diff
@@ -518,9 +518,35 @@ cluster.addCapacity('BottlerocketNodes', {
...
-Please note Bottlerocket does not allow to customize bootstrap options and `bootstrapOptions` properties is not supported when you create the `Bottlerocket` capacity. 
+Please note Bottlerocket does not allow to customize bootstrap options and `bootstrapOptions` properties is not supported when you create the `Bottlerocket` capacity.
 
+### Service Accounts
 
+With services account you can provide Kubernetes Pods access to AWS resources.
+
+```ts
+// add service account
+const serviceAccount = cluster.addServiceAccount('MyServiceAccount'));
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/README.md:529]:
```suggestion
const serviceAccount = cluster.addServiceAccount('MyServiceAccount');
```

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 25]
```diff
@@ -617,6 +620,45 @@ export class Cluster extends Resource implements ICluster {
...
+    if (!this._clusterResource) {
+      throw new Error('unable to obtain OpenID Connect issuer URL. Cluster must be kubectl-enabled');
+    }
+
+    return this._clusterResource.attrOpenIdConnectIssuerUrl;
+  }
+
+  /**
+   * Lazily creates the OpenIdConnectProvider resource, which links the cluster to AWS IAM.
+   */
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:640]: Focus on what this is from a user perspective and less about the implementation:

```suggestion
  /**
   * An `OpenIdConnectProvider` resource associated with this cluster, and which can be used 
   * to link this cluster to AWS IAM.
   * 
   * A provider will only be defined if this property is accessed (lazy initialization).
   */
```

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+import { Construct } from '@aws-cdk/core';
+import { Cluster } from './cluster';
+
+/**
+ * Service Account
+ */
+export interface ServiceAccountOptions {
+  /**
+   * The name of the service account.
+   * @default If no name is given, it will use the id of the resource.
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:11]: We use "`-`" to represent that the default is "behavioral"

```suggestion
   * @default - If no name is given, it will use the id of the resource.
```

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+   */
+  readonly name?: string;
+
+  /**
+   * The namespace of the service account.
+   * @default default
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:17]:
```suggestion
   * @default "default"
```

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+   */
+  readonly namespace?: string;
+}
+
+/**
+ * Service Account
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:23]:
```suggestion
 * Properties for defining service accounts
```

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 29]
```diff
@@ -0,0 +1,91 @@
+import { IPrincipal, IRole, OpenIdConnectPrincipal, PolicyStatement, PrincipalPolicyFragment, Role  } from '@aws-cdk/aws-iam';
+import { Construct } from '@aws-cdk/core';
+import { Cluster } from './cluster';
+
+/**
+ * Service Account
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:6]:
```suggestion
 * Options for `ServiceAccount`
```

File: packages/@aws-cdk/aws-iam/lib/principals.ts [hunk 30]
```diff
@@ -417,6 +418,55 @@ export class FederatedPrincipal extends PrincipalBase {
...
+
+  public toString() {
+    return `WebIdentityPrincipal(${this.federated})`;
+  }
+}
+
+/**
+ * A principal that represents a federated identity provider as from a OpenID Connect provider.
+ */
+export class OpenIdConnectPrincipal extends WebIdentityPrincipal {
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/lib/principals.ts:449]: Very nice... Mind referencing this (and `WebIdentityPrincipal`) in the README for the IAM library?

File: packages/@aws-cdk/aws-iam/test/principals.test.ts [hunk 31]
```diff
@@ -101,4 +101,18 @@ test('can have multiple principals the same conditions in the same statement', (
       }),
     ],
   }));
+});
+
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/test/principals.test.ts:105]: add a test for `WebIdentityPrincipal`

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+  /**
+   * The namespace where the service account is located in.
+   */
+  public readonly serviceAccountNamespace: string;
+
+  constructor(scope: Construct, id: string, props: ServiceAccountProps) {
+    super(scope, id);
+
+    const { cluster } = props;
+    this.serviceAccountName = props.name ?? id;
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:61]: How unique do we need this name? Perhaps we should default to `this.node.uniqueId`?

Comment (vlesierse) [reply packages/@aws-cdk/aws-eks/lib/service-account.ts:61]: They should be unique within a namespace but I agree `this.node.uniqueid` would be better.

File: packages/@aws-cdk/aws-eks/README.md (hunk 24 above)
Comment (vlesierse) [reply packages/@aws-cdk/aws-eks/README.md]: Added a warning explaining the missing role conditions and the implication of this.

File: packages/@aws-cdk/aws-iam/README.md [hunk 33]
```diff
@@ -198,6 +198,27 @@ const principal = new iam.AccountPrincipal('123456789000')
...
+
+```ts
+const principal = new iam.WebIdentityPrincipal('cognito-identity.amazonaws.com')
+  .withConditions({
+    "StringEquals": { "cognito-identity.amazonaws.com:aud": "us-east-2:12345678-abcd-abcd-abcd-123456" },
+    "ForAnyValue:StringLike": {"cognito-identity.amazonaws.com:amr": "unauthenticated"}
+  });
+```
+
+The `OpenIdConnectPrincipal` class can be used as a principal used with a `OpenIdConnectProvider`, for example:
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/README.md:212]: Can you move this to the open is connect section please?

**Comment by vlesierse** [comment]
> You'll need to update the EKS integration tests by running `npm run integ integ.eks-cluster.js` as the error suggests.

Got it :+1: Took some time to get it run correctly on one of my AWS accounts.

**Comment by eladb** [comment]
@vlesierse updated PR title & message

**Comment by pahud** [comment]
Amazing! Thank you @vlesierse and @eladb for all the great work.
